### PR TITLE
Progress bar for cli

### PR DESCRIPTION
**Pull request by bmordan**
**go-ipfs**

**js-ipfs**

This is the progress bar from [https://github.com/visionmedia/node-progress](https://github.com/visionmedia/node-progress) which is being called in `js-ipfs-unixfs-engine` in the [builder](https://github.com/ipfs/js-ipfs-unixfs-engine/blob/master/src/builder/builder.js#L96).

I have also done a little house keeping in the `src/files/add` as the callbacks were getting a little unwieldy.

### CODE REVIEW

File: src/cli/commands/files/add.js [hunk 1]
```diff
@@ -158,34 +189,45 @@ module.exports = {
...
-
-    if (typeof ipfs.files.createAddPullStream === 'function') {
-      createAddStream = (cb) => {
-        cb(null, ipfs.files.createAddPullStream(options))
-      }
-    }
+    let list = []
+    let currentBytes = 0
+
+    async.waterfall([
```
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: lets stick with `const waterfall = require('async/waterfall)` please

**Comment by bmordan** [comment]
Locally `npm run lint` returned no errors

Travis reported

```
/home/travis/build/ipfs/js-ipfs/examples/traverse-ipld-graphs/git.js
  16:3   warning  Unexpected console statement                 no-console
  39:7   warning  Unexpected console statement                 no-console
  50:22  error    Missing space before function parentheses    space-before-function-paren
  56:46  error    Expected space or tab after '//' in comment  spaced-comment
  60:9   warning  Unexpected console statement                 no-console
  61:9   warning  Unexpected console statement                 no-console
  62:9   warning  Unexpected console statement                 no-console
  65:1   error    More than 1 blank line not allowed           no-multiple-empty-lines
```

I'm about to commit some additional changes to pass the linting.

**Comment by diasdavid** [comment]
Also, does this work over ipfs-api? (i.e with a daemon on?)

**Comment by bmordan** [comment]
-[x] work with the daemon on (through js-ipfs-api) https://github.com/ipfs/js-ipfs-api/pull/596
-[ ] tests

**Comment by bmordan** [comment]
@dignifiedquire @diasdavid

## Testing the progress bar

There is now:

* a unit test for the progress bar in `js-ipfs`.
* tests in `js-ipfs-unixfs-engine`
  *to test that a function passed with `options.progress` to the `Importer` is called with the right payload*
* tests in `js-ipfs-api` to test that the progress function is called with the daemon running

What I tried to do yesterday was hook into progress.stdout and verify that the progress bar was being printed to the terminal.
I did manage this - is very sketchy and I decided to not submit it.

When the tests run progress.stdout is passed to the progress bar [library](https://github.com/visionmedia/node-progress) but `stdout.isTTY` (is node a text terminal) is false so the library returns early and doesn't output anything.

To get the progress bar to output in tests means you have to highjack stdout in the tests and stub out the following:

* isTTY = true
* columns = 80
* cursorTo = function () {}
* clearLine = function () {}

Then the progress bar is piped to `out` most of the `files add` tests break because what is written to stdout has changed. Before starting to re-created a `clearLine` function to conditionally clear the progress bar once it is complete one has to stop and ask one's self. What am I doing here? This is a bit of a rabbit hole and I think the tests we have for the progress bar are sufficient.

However I can see the testing coverage has now fallen 😞

**Comment by diasdavid** [comment]
Now done through https://github.com/ipfs/js-ipfs/pull/1036

**Pull request closed without merging**